### PR TITLE
[Large] Priority and automatic email systems for admission

### PR DESCRIPTION
**Pull request by evgiz**
This PR changes the admission system to enable automatic rejection emails.

- Applications now have priority (for cooperation between groups) and status (accepted/rejected) for automatic emails
- Admissions manager have better tools to keep track of admission status
- Tools for sending automatic rejection emails

### Some screenshots to explain systems

Interview priority system:
Screenshot 2020-11-21 at 19 16 01

Admissions admin overview:
Screenshot 2020-12-17 at 16 15 15

Review rejection recipients:
Screenshot 2020-12-17 at 16 16 52

### CODE REVIEW

File: app/controllers/admissions_admin/admissions_controller.rb [hunk 1]
```diff
@@ -16,6 +16,139 @@ def show
...
+        subject: params[:subject],
+        intro: params[:intro],
+        content: params[:content]
+    }
+
+    @success = []
+    @failure = []
+    @errors = []
+
+    @recipients.each do |r|
```
Comment (evgiz) [inline app/controllers/admissions_admin/admissions_controller.rb:126]: This function sends the actual rejection emails. Important to get this part right!

File: app/controllers/admissions_admin/admissions_controller.rb [hunk 2]
```diff
@@ -99,6 +231,37 @@ def admission_params
...
+  # Calculate who should receive rejection email
+  # Excludes those who already received an email
+  def calculate_rejection_recipients
+    rejected = automatically_rejected_applicants
+    already_sent = @admission.rejection_emails.map { |r| r.applicant_id }
+    # Select all not already sent
+    rejected.select { |r| not already_sent.include?(r.id) }
+  end
+
+  # Calculates who are marked to receive automatic rejection email
```
Comment (evgiz) [inline app/controllers/admissions_admin/admissions_controller.rb:243]: This function and "calculate_rejection_recipients" are very important to verify for correctness!

File: app/controllers/admissions_admin/admissions_controller.rb [hunk 3]
```diff
@@ -16,6 +16,139 @@ def show
...
+      begin
+        rej.save!
+        AdmissionRejectionMailer.send_rejection_email(r, @template).deliver
+        @success.append(r)
+      rescue Error => e
+        rej.delete!
+        @failure.append(r)
+        @errors.append(e)
+      end
+    end
```
Comment (evgiz) [inline app/controllers/admissions_admin/admissions_controller.rb:143]: This loop sends the actual rejection emails. Important to get this part right!
 Write

File: app/controllers/admissions_admin/admissions_controller.rb [hunk 4]
```diff
@@ -99,6 +231,37 @@ def admission_params
...
+
+    # Convert to unique applicants
+    rejected_somewhere = rejected_somewhere.flatten.map { |x| x.applicant }.uniq
+    accepted_somewhere = accepted_somewhere.flatten.map { |x| x.applicant }.uniq
+
+    # Return those rejected somewhere if not accepted anywhere
+    rejected_somewhere - accepted_somewhere
+  end
+
```
Comment (evgiz) [inline app/controllers/admissions_admin/admissions_controller.rb:263]: This function and "calculate_rejection_recipients" are very important to verify for correctness!

File: app/assets/javascripts/applicants/admissions_admin_email.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+$(function() {
+    $('input').keyup(function (event) {
+        var intro = $('input[name="introduction"]').val()
+        var subj = $('input[name="subject"]').val()
+        $("#preview-subject").text(subj)
+        $("#preview-intro").text(intro+" Pernille,")
```
Comment (eirikvaa) [inline app/assets/javascripts/applicants/admissions_admin_email.js]: Should probably change the default name to something like `<Name>`.

File: config/locales/views/interviews/no.yml [hunk 6]
```diff
@@ -1,11 +1,11 @@
 'no':
   interviews:
-    administer_interviews: "Planlegg intervjuer"
+    administer_interviews: "Administrer søknader"
```
Comment (eirikvaa) [inline config/locales/views/interviews/no.yml:3]: `administer_applications`?

File: app/assets/javascripts/applicants/admissions_admin_email.js (hunk 5 above)
Comment (evgiz) [reply app/assets/javascripts/applicants/admissions_admin_email.js]: True, this was really just a silly easter egg for the current admission leader. Should probably do this to make it clear that the name is automatically set!

Comment (eirikvaa) [reply app/assets/javascripts/applicants/admissions_admin_email.js]: Ah, didn't realize it was an easter egg 😅 That changes things!

File: app/views/admissions_admin/jobs/show.html.haml [hunk 7]
```diff
@@ -1,46 +1,148 @@
...
 
+  = render_async admissions_admin_admission_group_job_show_unprocessed_path(job_id: @job) do
+    .samf-container.flex-col.p-3.text-Align.mb-3
+      %h2= t("interviews.loading_title")
+      %p.mb-2{style: "color: #666;"}= t("interviews.loading_description")
+      .loading-indicator.large.center
 
-= render 'jobs_show', job_application_groupings: [@job.job_applications], show_job_titles: false
+- else
+  %p= t("interviews.no_unprocessed_applicants")
```
Comment (eirikvaa) [inline app/views/admissions_admin/jobs/show.html.haml:39]: Would be better if we could show applications in some way, even though they have been processed. Some hidden field that could be expanded, perhaps?

**Comment by eirikvaa** [comment]
@evgiz So now it seems that if you get an offer for a job and decline it, you _still_ get a rejection email? Is that really necessary?

File: app/models/job.rb [hunk 8]
```diff
@@ -66,6 +66,65 @@ def job_applications_without_interviews
     job_applications - job_applications_with_interviews
   end
 
+  def total_processed_applications
```
Comment (eirikvaa) [inline app/models/job.rb]: Do you need these `total_` methods when you have counterpart methods that just return the objects?

**Review by eirikvaa — APPROVED** [review]
This looks good to me. Pages look visually good. The overview page under admissions admin seems to work as expected, and changes made to interviews/applicants seem to be reflected as expected on this page. Ran through the rejection sending flow, and my browser went bananas as it opened 300 tabs :-)

The logic that calculates the rejection recipients and sending the emails also seem to be good. Don't see any obvious faults in them.

**Pull request merged**